components-java-sdk/src/main/java/io/dapr/components/domain/state/TransactionableOperation.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.domain.state;
+
+/**
+ * This interface act as a marker for operations that can be used with Transactions.
+ *
+ * <p>This interface is being used as a type-safe alternative to representing the enum/oneof
+ * used by {@link dapr.proto.components.v1.State.TransactionalStateOperation}</p>
+ */
+public interface TransactionableOperation {
```

Comment:
I like the fact that this minimizes the number of types (i.e. doesn't require a wrapper "operation" type).  The concern I have is that it might be more difficult for developers to understand what possible types to expect, as there's no finite list of operation types to map to request types.  The developer must find all implementors of `TransactionableOperation`.  It would be easy to miss new times added over time (however likely that might be).

While I also dislike the oneof/union type as defined in the `.proto`, having a `TransactionableOperation` abstract type with a similar enum offers developers a compile-type checkpoint that all known operations types have been accounted for (even if the mapping from operation type to operation request class type is still a soft contract--that, at least, can be easily documented).

For example, in .NET, switch expressions could be used to handle such cases, and the compiler would automatically warn/error if not all cases were handled.

```csharp
public enum StateStoreTransactOperationType
{
    Delete,
    Set
}

public abstract record StateStoreTransactOperation(StateStoreTransactOperationType Type)
{
}

public sealed record StateStoreTransactOperation<T>(StateStoreTransactOperationType Type, T Request)
    : StateStoreTransactOperation(Type)
{
}

StateStoreTransactOperation operation = ...

return operation.Type switch
{
  case TransactionableOperationType.Set => ProcessSetRequest(((StateStoreTransactOperation<SetRequest>)operation).Request)
  case TransactionableOperationType.Delete => ProcessDeleteRequest(((StateStoreTransactOperation<DeleteRequest>)operation).Request)
}
```

Adding just the type enum as a property of `TransactionableOperation` might be an improvement, sort of a compromise between the two.  The downside is that marking `SetRequest` as `Transactionable` might be confusing when that request is used outside of the transaction request (i.e. from a normal set request).  That is, the  `SetRequest` is not inherently transactionable and doesn't, itself, contribute to the transaction.  Whether it's handled in a transactional way is solely based on context.  But, I'm probably overthinking it.

Reply:
Representing this `oneof` is a bit of a pickle, isn't it? 🤣

If we had union types in Java (and C#) this would not be a bit of an issue. But we don't and I think the best at this point is to do some sort of python2 solution (runtime dependent).

In Java17 with `--enable-preview ` has a limited form of pattern matching that would allow for something as _elegant_ as:

```java
converted = fromNetworkTransactions.getOperations.stream()
  .map( op -> switch (op) {
      State.SetRequest setOp -> SetRequest(setOp)
      State.DeleteRequest deleteOp -> DeleteRequest(deleteOp)
      default -> throw WhatIsThisThatYouSentDownTheWireThisIsNotAnExpectedTypeHere("ok?");
    })
  .toList();
```
A bit better. But in the end this is still runtime based.

Granted, something similar could still be done in Java8-Java17 using a series of  `if`+`instanceOf` clauses.

---

Back to your point We could use `enums` as you mentioned. But I don't think it is making it easier on developers honestly.

It does document it clearly. We could have JavaDoc in that interface stating "this interface is only used by these 2 types: Set and Delete" but that's fragile.

Java, AFAIK, won't warn if your are not being `exhaustive` in switch on enum statements.

In Java, aside runtime/reflection, we could have a pair-like class to encapsulate the enum+instance but I wonder if this is really making it easier on the dev. 

What I considered as solution at a point was to add methods such as `addDeleteOperation` and `addSetOperation` to `TransactionalStateRequest` or to is (to be created) builder class.

How about defining a helper Visitor class/method? So we remove from the developer the need to even think about the switch and parsing of that list?

```java
/**
 * Iterates over each item in the list in orderm invoking the appropriate visitor functor/consumer
 * in that order, depending on its type.
 */
void forEachOperation(Consumer<DeleteRequest> deleteVisitor, Consumer<SetRequest> setVisitor) {
  for (var operation : this.operations() {
    if (op isInstanceOf DeleteRequest deleteRequest) {
      deleteVisitor.apply(deleteRequest)
    }
    else if (op isInstanceOf SetRequest setRequest) {
      setVisitor.apply(setRequest)
    } else {
      throw AnAppropriateRunTimeLikeExceptionHere();
    }
  };
}
```

An user could use this as

Reply:
The visitor pattern is an appealing one.  The parameters make it obvious what request types participate in the transaction, and help ensure they're all accounted for.  Practically it might be a little awkward (at least in .NET) as the sequential iteration would likely require both synchronous and asynchronous flavors, something like:

```csharp
public void VisitOperations(Action<DeleteRequest> onDeleteRequest, Action<SetRequest> onSetRequest);
public Task VisitOperationsAsync(Func<DeleteRequest, Task> onDeleteRequest, Func<SetRequest, Task> onSetRequest);
```

Reply:
As a variation, perhaps still have an `TransactionOperation` type so that you expose as a list from the root request (allowing normal iteration), but have that type not directly expose the underlying request in favor of a `Visit` method.

```csharp
public class TransactionOperation
{
  public Visit(Action<DeleteRequest> onDeleteRequest, Action<SetRequest> onSetRequest);
  public Visit<T>(Func<DeleteRequest, T> onDeleteRequest, Func<SetRequest, T> onSetRequest);
}

public record TransactionRequest(IReadOnlyList<TransactionOperation> Operations)
{
}
```

This would allow something like:

```csharp
foreach (var operation in request.Operations)
{
  await operation.Visit(
   deleteRequest => PerformDelete(deleteRequest),
   setRequest => PerformSet(setRequest));
}
```

In .NET, you'd still need both a void and non-void `Visit` method--oh, how I wish `void` was itself a type, so there's still a little bit of awkwardness there, but it at least removes iteration from the SDK's concerns.  (.NET guidelines typically frown on `.ForEach()` style methods.)

Reply:
I added this method in the last commit.

Reply:
I missed your two comments when I pushed the updated commit. But since the PR is not merged, let me address them.

Do we would we really need sync and async variations of this visitor? Or is just common practice doing this in .Net?

I am thinking on how this is used: it is consumed as a whole, synchronously, to translate the this transaction request into the underlying state store request. There might be StateStores that would allow each individual request to be issued o their own and async, but I can't think of one that would allow for such request to be issued concurrently.

Reply:
It's less about asynchrony in the sense of running in parallel and more a matter of potentially needing to perform an asynchronous action for each operation in sequence.  Unfortunately, C# doesn't have a "unit" type that would allow a single generic method to either return `void` for a synchronous visit or `Task` for an asynchronous visit.  You could start with one method that always returns a value to enable the latter and, as a bit of an ugly hack, the developer could just return a random value if they don't care about the result in order to achieve the former.